### ISSUE TITLE
'Incomplete credentials' error returned, though credentials are provided

### ISSUE THREAD

**Issue by samtuke** [phpList/rest-api#57]
### Short problem description
The following JSON request is sent but 'incomplete credentials' are reported:

```JSON
{
    "loginName": "admin",
    "password": "password"
}
```
Using a local PHP development server (http://127.0.0.1:8000/api/v2/sessions), with .htaccess in root dir removed (to avoid permission issues).

### Steps to reproduce
1. Send a login request with above JSON using Postman Chrome app
2. Observe reported error

### Expected behaviour
Session is created, success status returned

### Actual behaviour
Error reported

#### Package version
Installed base-distribution via composer
#### PHP and Composer version
PHP 7.1.11
Composer version 1.5.2 2017-09-11 16:59:25

**Comment by oliverklee** [comment]
I'll test this and find out what is missing in the integration tests.

**Comment by oliverklee** [comment]
Using the development server for the `rest-api` package (`bin/console server:run -d web/`), I get a ´401 not authorized´ response (which is correct). These are the parameters I've used in Postman:

POST
http://127.0.0.1:8000/api/v2/sessions
Header: Content-Type application/json
Body: 
```json
{
    "code": 401,
    "message": "Not authorized"
}
```

Does the problem also occur when you're using the development server?

(I'll also retest using the PhpStorm REST tools and the ´base-distribution´ package in a minute.

**Comment by oliverklee** [comment]
I get the same results with the `base-distribution` package.

**Comment by oliverklee** [comment]
The same with the REST client in PhpStorm.

**Comment by oliverklee** [comment]
@samtuke Are you using PHP's built-in web server (in development context) or Apache (in production context)? And if you're using Apache, does the problem also occur if you're using the built-in web server instead?

**Comment by samtuke** [comment]
I believe I tried it with both and experienced the same outcome

**Comment by oliverklee** [comment]
Could you please re-try? Maybe there was an error in the JSON request data? I currently cannot reproduce the problem.

**Comment by samtuke** [comment]
Following a `composer update` and copy & paste of login credentials from the docs, I get the same error when running a local development server:
```json
{
    "loginName": "admin",
    "password": "eetIc/Gropvoc1"
}
```
![postman_340](https://user-images.githubusercontent.com/695422/33528021-e2b3e5a8-d85a-11e7-98f9-5b2abd75637c.png)

**Comment by oliverklee** [comment]
I've used the standalone version of Postman for this (as the Chrome app has been discontinued). Could you please retest with the stand-alone Postman or the PhpStorm REST tools to exclude any problems with the Chrome app?

**Comment by oliverklee** [comment]
Oh, and in your screenshot, the content type (the drop-down right next to the type radio buttons (raw, binary etc.) is not set to "JSON (application/JSON)". Could you please set that as well?

**Comment by samtuke** [comment]
@oliverklee Changing the content type fixed it -- well spotted! It seems that if a request is provided, but not with JSON headers, then it returns an 'Incomplete credentials' error, as opposed to 'empty JSON data', which is returned if either empty JSON or text is submitted. Two potential improvements would be:

- If an empty request is submitted with text (not JSON) headers, then respond with error 'JSON missing' or similar
- If non-JSON is submitted, respond with 'Invalid request' or similar, instead of 'Incomplete credentials', as this implies that the request was JSON but the data incomplete.

**Comment by oliverklee** [comment]
Yes, we should definitely return an "invalid request" status with a helpful error message in that case. What exact Content-type header did you use?

**Comment by samtuke** [comment]
@oliverklee it looks like no content-type was set at all:

```
POST /api/v2/sessions HTTP/1.1
Host: 127.0.0.1:8000
Cache-Control: no-cache
Postman-Token: 83ca8f5e-6771-c158-0f6a-9fae28441756

{
    "loginName": "admin",
    "password": "phplist"
}
```

**Comment by littlepackage** [comment]
Three years later and I am also getting the "incomplete credentials" 400 error with this API. This one is really tricky to set up, to begin with the .htaccess in the lists folder was not going to work on my server (Litespeed). Had to do:

```
<IfModule mod_rewrite.c>
    Options -MultiViews
    RewriteEngine On
	RewriteRule api/v2.*? api.php [L]
</IfModule>
```

Instead. Then at least the API could be reached. I'm using wp_remote_post() (Wordpress) and just trying to get credentials. Body looks like:

```
$body = wp_json_encode([
    form_params' => [
         loginName' => $u,
         password' => $p,
    ],
]);
```

This body is based off [the example PHPList team has linked to](https://github.com/phpList/RestAPIclientSimpleExample/blob/master/RESTAPIphpListClient.php).

```
$headers = [
    'Content-Type' => 'application/json'
];
$response = wp_remote_post( 
    'https://mywbsite.com/lists/api/v2/sessions', 
        array(
            'method'		=> 'POST',
            'httpversion' 	=> '1.1',
            'headers'     	=> $headers,
            'body'        	        => $body,
	    'timeout'     	=> 15,
        )
);
```

As you can see Content Type is there and it is confirmed in the API response. Have tried things in all sorts of arrangements and am pulling my hair at this point. Version 3.6.2 upgrading right now to 3.6.3... Any ideas?
